src/fdcache.cpp

```diff
@@ -754,7 +754,8 @@ bool FdManager::Close(FdEntity* ent, int fd)
 bool FdManager::ChangeEntityToTempPath(FdEntity* ent, const char* path)
 {
     const std::lock_guard<std::mutex> lock(FdManager::except_entmap_lock);
-    except_fent[path] = ent;
+    except_fent[path] = std::shared_ptr<FdEntity>(ent);
```

Comment:
Do we have any kind of test that demonstrate the existing double delete?  I worry that there will be two separate `std::shared_ptr` each with reference count = 1 that will continue to double-delete the raw pointer.  For this to work we should expect a single `std::shared_ptr` with reference count = 2.  The crashes we are seeing are demonstrating that s3fs has a memory ownership misunderstanding -- where are we creating the second reference to the same raw pointer?  Ideally whatever allocates the `FdEntity` should wrap the raw pointer in an owned pointer (either `std::unique_ptr` or `std::shared_ptr` and all reference should go through that.

Reply:
@ggtakec  
The current issue is that two shared_ptr are referencing the same raw pointer. I checked out your PR for verification, and the problem still persists.
When the deadlock occurs, the following content is printed in the logs:
> s3fs: [WAN] fdcache.cpp:UpdateEntityToTempPath(787): For some reason the FdEntity pointer(for /200M) is not found in the fent map. Recovery procedures are being performed, but the cause needs to be identified.

## To reporduce the deadlock, you could fellow these steps
1. modify small-integration-test.sh

```
diff --git a/test/small-integration-test.sh b/test/small-integration-test.sh
index 6a1029f..5f22388 100755
--- a/test/small-integration-test.sh
+++ b/test/small-integration-test.sh
@@ -48,18 +48,18 @@ export CACHE_DIR
 export ENSURE_DISKFREE_SIZE 
 if [ -n "${ALL_TESTS}" ]; then
     FLAGS=(
-        "use_cache=${CACHE_DIR} -o ensure_diskfree=${ENSURE_DISKFREE_SIZE} -o fake_diskfree=${FAKE_FREE_DISK_SIZE} -o use_xattr -o update_parent_dir_stat"
-        enable_content_md5
-        disable_noobj_cache
-        "max_stat_cache_size=100"
-        nocopyapi
-        nomultipart
-        sigv2
-        sigv4
-        "singlepart_copy_limit=10"  # limit size to exercise multipart code paths
-        #use_sse  # TODO: S3Proxy does not support SSE
-        #use_sse=custom:/tmp/ssekey  # TODO: S3Proxy does not support SSE
-        "use_cache=${CACHE_DIR} -o ensure_diskfree=${ENSURE_DISKFREE_SIZE} -o fake_diskfree=${FAKE_FREE_DISK_SIZE} -o streamupload"
+        "ensure_diskfree=${ENSURE_DISKFREE_SIZE} -o fake_diskfree=${FAKE_FREE_DISK_SIZE}"
+        # enable_content_md5
+        # disable_noobj_cache
+        # "max_stat_cache_size=100"
+        # nocopyapi
+        # nomultipart
+        # sigv2
+        # sigv4
+        # "singlepart_copy_limit=10"  # limit size to exercise multipart code paths
+        # #use_sse  # TODO: S3Proxy does not support SSE
+        # #use_sse=custom:/tmp/ssekey  # TODO: S3Proxy does not support SSE
+        # "use_cache=${CACHE_DIR} -o ensure_diskfree=${ENSURE_DISKFREE_SIZE} -o fake_diskfree=${FAKE_FREE_DISK_SIZE} -o streamupload"
     )
 else
     FLAGS=(
```

2.  start s3fs with following ENV
```
ALL_TESTS=1 INTERACT=1 DBGLEVEL=debug ./small-integration-test.sh
```

3. cp a 180MB file in s3fs dir

Reply:
I’m sorry for this code still has problems, so I will update it after checking again.

@gaul
I will check if I can prepare test cases etc.
(Additinal comment: probably my previous comment(the reason is rename) was wrong.)
However, I think it is possible that the same FdEntity is assigned to multiple keys depending on the timing of processing.
As reported by @VVoidV, I think it occurs when the file cache is unused and disk space is insufficient.
(And I think the problem is that the FdEntity is deadlocked when it is closed. The fix in #2478 was insufficient.)
When the cache space is insufficient, the fdentity map's key is registered as a string that is not a file path, and then when the capacity runs out again, I think that there is a possibility that the same FdEntity will be registered to multiple keys with a later call to `UpdateEntityToTempPath()`.
In any case, the current this PR code still needs to be revised, so please review it again after the revisions are made.

@VVoidV
Thank you. I have confirmed the deadlock.
I'll update this PR code, please wait a little longer.

Reply:
@gaul @VVoidV 
This source code has been slightly updated. Please review again with the new code.